resolver/dns/dns_resolver_test.go

```diff
@@ -1126,3 +1142,65 @@ func TestCustomAuthority(t *testing.T) {
 		}
 	}
 }
+
+// TestRateLimitedResolve exercises the rate limit enforced on re-resolution
+// requests. It sets the re-resolution rate to a small value and repeatedly
+// calls ResolveNow() and ensures only the expected number of resolution
+// requests are made.
+func TestRateLimitedResolve(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	dc := replaceDNSResRate(500 * time.Millisecond)
+	defer dc()
+
+	// Create a new testResolver{} for this test because we want the exact count
+	// of the number of times the resolver was invoked.
+	nc := replaceNetFunc()
+	defer nc()
+
+	target := "foo.bar.com"
+	b := NewBuilder()
+	cc := &testClientConn{target: target}
+	r, err := b.Build(resolver.Target{Endpoint: target}, cc, resolver.BuildOption{})
+	if err != nil {
+		t.Fatalf("resolver.Build() returned error: %v\n", err)
+	}
+	defer r.Close()
+
+	dnsR, ok := r.(*dnsResolver)
+	if !ok {
+		t.Fatalf("resolver.Build() returned unexpected type: %T\n", dnsR)
+	}
+	tr, ok := dnsR.resolver.(*testResolver)
+	if !ok {
+		t.Fatalf("delegate resolver returned unexpected type: %T\n", tr)
+	}
+
+	// We set the re-resolution rate to 500ms at the start of this test. After
+	// this loop, we expect to see two and only two resolution requests making it
+	// past the rate limiter.
+	start := time.Now()
+	end := start.Add(750 * time.Millisecond)
+	for time.Now().Before(end) {
+		r.ResolveNow(resolver.ResolveNowOption{})
+	}
```

Comment:
Maybe sleep for 1ms here to avoid a ridiculous amount of ResolveNow calls?

Also, count the calls you do make and t.Fatal() if you didn't do more than 2.

Reply:
Done.